### ISSUE TITLE
Extra function signatures in unity source

### ISSUE THREAD

**Issue by austinglaser** [ThrowTheSwitch/Unity#191]
This is a style, rather than functional, issue.

There are many instances similar to the following (line 112 in unity.c):

```
void UnityPrintLen(const char* string, const _UU32 length);
void UnityPrintLen(const char* string, const _UU32 length)
{
    /* Function body */
}
```

Forward declarations have their place, but in this case it adds no real extra value while using an extra line. I suggest removing the forward declarations, especially since it's inconsistent within the file.

If needed, I can submit a pull request with the patch.

**Comment by mvandervoord** [comment]
Sorry, the C99 and later standards require forward declarations. We have therefore included them to avoid compiler warnings and errors.

We COULD organize them at the top of the file for convenience, but there's no real reason to have them in any particular place at the moment.

**Comment by zanedp** [comment]
What compiler has a problem here? I can't get gcc nor clang to issue a warning the declaration is removed with -std=c99. I've also never had IAR issue a warning for this. Where does C99 say you have to declare a function before defining it? I see "remove implicit function declaration", but I don't see anywhere that explicit declarations are always required. It must be _at least_ declared before it's called, but it doesn't need to be declared before it's defined. A definition before it's called "counts" as a declaration because the definition includes a "declarator".

**Comment by mvandervoord** [comment]
Hm. You are correct about that. C99+ only requires that a declaration and/or definition appear before the function is used.

These lines were not always there. I can distinctly remember merging the change that put them there because someone had a problem... but I kept reaching dead ends when I ran a blame to find the commit that made that happen because we have since fixed whitespace and line endings. I was hoping to go back to the original commit and see why we thought it was necessary. Perhaps it was only necessary on one particular function for some reason and the others were just "fixed" at the same time.

I'll do a bit more digging. :)

**Comment by mjago** [comment]
I just did some digging too and the relevant commit appears to be:
[https://github.com/ThrowTheSwitch/Unity/commit/77af37add04a5899771f4a163650443e9f8fb24f](commit) 

```
Commit: 77af37add04a5899771f4a163650443e9f8fb24f
Author: ThingamaByte, LLC <mark@thingamabyte.com>
Date:   Thu Jun 18 17:16:01 2015 -0400 (11 months ago)
Path:   src/unity.c
Match:  3 of 12

 catch optional fixtures plugin up to mainline unity (somewhat. like fixing line endings)

```

So it doesn't really look to have been a targeted change.

**Comment by mvandervoord** [comment]
Thanks, Martyn.

So that commit is me (no surprise... I thought I remembered doing the
merge)... but I really *think* it came from somewhere else first. Not sure,
though.

Anyway, I think I am onboard with stripping those back out. If we can
eliminate another warning, that is always a good thing.

**Comment by jsalling** [comment]
I also looked into this. I found this:

These extra declarations are required when compiling with `-Wmissing-prototypes`. used in:
* Example 2 that comes with Unity
* the Fixture Makefile

Note: `testunity.c` will not build with `missing-prototypes` set to `error`. It does not include a header or prototypes for any functions. The test build scans this file directly to produce the Runner.c file. As such, Unity's tests will not mesh with this flag, and it is not enabled there.

I'm sensing that the examples should be added to the travis CI build, so they stay up to date. I can create a separate issue for that.

* This particular function prototype, `UnityPrintLen`, might move to the internals header, since it could be used in `testunity.c` like `UnityPrint`.
* Functions that are `static` will compile without the extra function signature.
* Others might need a case by case decision
